project/src/main/java/com/google/sps/data/VisionLibrary.java

```diff
@@ -0,0 +1,330 @@
+        return imagesService.getServingUrl(options);
+      }
+    } catch (Exception e) {
+      return "/serveBlobstoreImage?blobKey=" + blobKey.getKeyString();
+    }
+  }
+
+  /**
+   * Blobstore stores files as binary data. This function retrieves the binary data stored at the
+   * BlobKey parameter.
+   */
+  public static byte[] getBlobBytes(BlobstoreService blobstoreService, BlobKey blobKey) throws IOException {
+    ByteArrayOutputStream outputBytes = new ByteArrayOutputStream();
+
+    int fetchSize = BlobstoreService.MAX_BLOB_FETCH_SIZE;
+    long currentByteIndex = 0;
+    boolean continueReading = true;
+    while (continueReading) {
+      // end index is inclusive, so we have to subtract 1 to get fetchSize bytes
+      byte[] b =
+          blobstoreService.fetchData(blobKey, currentByteIndex, currentByteIndex + fetchSize - 1);
+      outputBytes.write(b);
+
+      // if we read fewer bytes than we requested, then we reached the end
+      if (b.length < fetchSize) {
+        continueReading = false;
+      }
+
+      currentByteIndex += fetchSize;
+    }
+
+    return outputBytes.toByteArray();
+  }
+  
+  public static String getNearestColor(Color color) {
+    Color[] constantColors = new Color[] { Color.black, Color.blue, Color.cyan, Color.gray, Color.green, Color.magenta, Color.orange, Color.pink, Color.red, Color.white, Color.yellow };
+    Color nearestColor = null;
+    Integer nearestDistance = new Integer(Integer.MAX_VALUE);
+
+    for (Color constantColor : constantColors) {
+      int newDistance = colorDistance(constantColor, color);
+      if (newDistance < nearestDistance) {
+        nearestColor = constantColor;
+        nearestDistance = newDistance;
+      }
+    }
+
+    return getNearestColorString(nearestColor);
+  }
+
+  // https://stackoverflow.com/questions/6334311/whats-the-best-way-to-round-a-color-object-to-the-nearest-color-constant
+  // This link explaines the reasoning behind the magic numbers present in this function.
+  // the main idea is that the closest color visually is not necessarily the closest
+  // color in terms of a strict euclidean distance.
+  public static int colorDistance(Color c1, Color c2) {
+    // int red1 = c1.getRed();
+    // int red2 = c2.getRed();
+    // int rmean = (red1 + red2) >> 1;
+    // int r = red1 - red2;
+    // int g = c1.getGreen() - c2.getGreen();
+    // int b = c1.getBlue() - c2.getBlue();
+    // return (((512+rmean)*r*r)>>8) + 4*g*g + (((767-rmean)*b*b)>>8);
+    int r = c1.getRed() - c2.getRed();
+    int g = c1.getGreen() - c2.getGreen();
+    int b = c1.getBlue() - c2.getBlue();
+    return r*r + g*g + b*b;
+  }  
+
+  public static Color convertToJavaColor(com.google.type.Color color) {
+    return new Color(Math.round(color.getRed()), Math.round(color.getGreen()), Math.round(color.getBlue()));
+  }
+
+  public static String getNearestColorString(Color color) {
+    if (color.equals(Color.black)) return "black";
```

Comment:
You can replace this with a switch statement to make it a bit cleaner.

https://www.w3schools.com/java/java_switch.asp

Reply:
From my brief search online, it seems that case statements don't work except for primitive types such as strings? So I can't do something like because it doesn't know equality:
switch (color) {
    case Color.black {return "black"; break;}
    ...
}
Not exactly sure, but what would be a good way to approach this?

Reply:
Ah, you're right, sorry about that. I hadn't noticed the type of `color`.

In that case, I think what you have is already good.